pkg/cnb/buildpack_retriever.go

```diff
@@ -0,0 +1,112 @@
+package cnb
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/v1"
+
+	"github.com/pivotal/kpack/pkg/apis/experimental/v1alpha1"
+)
+
+type BuildpackClient interface {
+	Fetch(keychain authn.Keychain, repoName string) (v1.Image, error)
+}
+
+type BuildpackRetriever struct {
+	Keychain authn.Keychain
+	Client   BuildpackClient
+	Store    *v1alpha1.Store
+}
+
+type BuildpackNotFound error
+
+func buildpackNotFoundError(msg string) BuildpackNotFound {
+	return BuildpackNotFound(errors.New(msg))
+}
+
+func (b *BuildpackRetriever) FetchBuildpack(id, version string) (RemoteBuildpackInfo, error) {
+	var storeBuildpack *v1alpha1.StoreBuildpack = nil
+	for _, buildpack := range b.Store.Status.Buildpacks {
+		if buildpack.ID == id {
+			if version == "" || buildpack.Version == version {
+				storeBuildpack = &buildpack
+				break
+			}
+		}
+	}
+
+	if storeBuildpack == nil {
+		return RemoteBuildpackInfo{}, buildpackNotFoundError(fmt.Sprintf("could not find buildpack with id '%s' and version '%s'", id, version))
+	}
+
+	buildPackageImage, err := b.Client.Fetch(b.Keychain, storeBuildpack.BuildPackage.Image)
```

Comment:
Refetching each buildpackage over and over is really expensive for each reconciliation.

Reply:
Agreed. Let's discuss alternatives.